### ISSUE TITLE
Benchmarking should use minimum time, not mean time

### ISSUE THREAD

**Issue by oxinabox** [google-research/dex-lang#373]
[Chen, J. and Revels, J., 2016. Robust benchmarking in noisy environments. arXiv preprint arXiv:1608.04295.](https://arxiv.org/abs/1608.04295) justifies this more extensively (and gives a more complete procedure for benchmarking),
but it boil does to: 
_noise can only ever make things run slower, never faster._
as such `minimum` not `mean` or `median` is the best estimator of the true time.
Since the model is `true_time + noise` where `noise` is nonnegative, and thus the sample that is smallest must have the smallest error (since `true_time` doessn't chane between samples, only noise).

Technically there are other things to consider, like CPU branch prediction can become primed if you keep running the same code, but that has a different solution.
And maybe you are worried about how it preforms while running on a under realistic load, but that again seems like a different problem and not one for microbenchmarking.

**Comment by dan-zheng** [comment]
I wonder what do [google/benchmark](https://github.com/google/benchmark) (C++) and [google/swift-benchmark](https://github.com/google/swift-benchmark) (Swift) do?

---

Those benchmarking libraries (especially the first) are more robust and complete and have features like customizable statistics, "time cap so benchmarking slow functions doesn't run too many times", and more. We should definitely adapt some of those ideas.

**Comment by oxinabox** [comment]
Yeah, a more serious attempt at benchmarking should probably allow access to the raw times or at least ability to pass in estimantors.
but for the basic functionality, if you have to have 1 estimator, `minimum` is the one you want.

**Comment by oxinabox** [comment]
> I wonder what do google/benchmark (C++) and google/swift-benchmark (Swift) do?

I think they both report `mean`.

There is a nice piece: [Minimum Times Tend to Mislead When Benchmarking](https://tratt.net/laurie/blog/entries/minimum_times_tend_to_mislead_when_benchmarking.html).
Which does make a number of good points which basically boil down to benchmarking is hard because of nondeterminism and nondeterminism is way more common than you think.
But its not making a case for perfering `mean` over `minimum` so much as for thinking carefully when designing benchmarks and looking at histograms.
Maybe that is the answer, we should default to showing histograms, but single numbers feel so clean.
(also just showing histograms of normal sample are probably not enough to be doing good by that blog post, you also might need to do things like using a varient of different data for different samples).

**Comment by oxinabox** [comment]
Apparently there is a collection of arguments about why mean us better here
https://pyperf.readthedocs.io/en/latest/analyze.html#minimum-vs-average

Because pyperf uses mean for it's `timeit`, 
But cpython't `timeit` uses minimum?
